core/imds/src/main/java/software/amazon/awssdk/imds/internal/EndpointProvider.java

```diff
@@ -0,0 +1,168 @@
+        }
+
+        EndpointMode finalEndpointMode = resolveEndpointMode(endpointMode);
+        if (finalEndpointMode == EndpointMode.IPV4) {
+            return EC2_METADATA_SERVICE_URL_IPV4 ;
+        } else if (finalEndpointMode == EndpointMode.IPV6) {
+            return EC2_METADATA_SERVICE_URL_IPV6 ;
+        } else {
+            throw SdkClientException.create("Unknown endpoint mode: " + endpointMode);
+        }
+    }
+
+    public EndpointMode getEndpointMode() {
+
+        Optional<String> endpointMode = SdkSystemSetting.AWS_EC2_METADATA_SERVICE_ENDPOINT_MODE.getNonDefaultStringValue();
+        if (endpointMode.isPresent()) {
+            return EndpointMode.fromValue(endpointMode.get());
+        }
+
+        return configFileEndpointMode()
+            .orElseGet(() -> EndpointMode.fromValue(SdkSystemSetting.AWS_EC2_METADATA_SERVICE_ENDPOINT_MODE.defaultValue()));
+    }
+
+    public String getEndpointOverride() {
+
+        Optional<String> endpointOverride = SdkSystemSetting.AWS_EC2_METADATA_SERVICE_ENDPOINT.getNonDefaultStringValue();
+        if (endpointOverride.isPresent()) {
+            return endpointOverride.get();
+        }
+
+        Optional<String> configFileValue = configFileEndpointOverride();
+
+        return configFileValue.orElse(null);
+    }
+
+
+    public String resolveEndpoint(URI endpoint, EndpointMode endpointMode) {
+        String finalEndpoint = resolveFinalEndpoint(endpoint, endpointMode);
+        if (finalEndpoint.endsWith("/")) {
+            finalEndpoint = finalEndpoint.substring(0, finalEndpoint.length() - 1);
+        }
+        return finalEndpoint;
+    }
+
+    public EndpointMode resolveEndpointMode(EndpointMode endpointMode) {
+        if (endpointMode != null) {
+            return endpointMode;
+        }
+
+        return getEndpointMode();
+    }
+
+    public String resolveFinalEndpoint(URI endpoint, EndpointMode endpointMode) {
+
+        if (endpoint != null) {
+            return String.valueOf(endpoint);
+        }
+        return getEndpoint(endpointMode);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    private Optional<EndpointMode> configFileEndpointMode() {
+        return resolveProfile().flatMap(p -> p.property(ProfileProperty.EC2_METADATA_SERVICE_ENDPOINT_MODE))
+                               .map(EndpointMode::fromValue);
+    }
+
+    private Optional<String> configFileEndpointOverride() {
+        return resolveProfile().flatMap(p -> p.property(ProfileProperty.EC2_METADATA_SERVICE_ENDPOINT));
+    }
+
+    private Optional<Profile> resolveProfile() {
+        ProfileFile profileFileToUse = resolveProfileFile();
+        String profileNameToUse = resolveProfileName();
+
+        return profileFileToUse.profile(profileNameToUse);
+    }
+
+    private ProfileFile resolveProfileFile() {
+        if (profileFile != null) {
+            return profileFile.get();
+        }
+
+        return ProfileFile.defaultProfileFile();
```

Comment:
Should we store this value in a `Lazy<ProfileFile>` so that we don't re-read the file each time? Or is this cached somewhere else?

Reply:
The SEP did not have inputs regarding re-read nor caching of config file.